Separate functionalities into modules

xhaggi opened the pull request:
As discussed in https://github.com/wimdeblauwe/htmx-spring-boot-thymeleaf/discussions/63 and to push things forward a bit, this PR separates the core and Thymeleaf functionality into modules. Since this changes the whole project structure, I bumped the project to the next major version `3.0.0-SNAPSHOT`.

The following modules are now available:

* `htmx-spring-boot-parent`
* `htmx-spring-boot`
* `htmx-spring-boot-thymeleaf`

I have moved all tests to their corresponding modules and renamed the auto configuration classes to `HtmxMvcAutoConfiguration` and `HtmxThymeleafAutoConfiguration` so that they are following the naming convention. I also moved the `org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc.imports` configuration into the test scope.

wimdeblauwe commented on spring-boot-htmx-thymeleaf/pom.xml:
```diff
@@ -0,0 +1,64 @@
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-configuration-processor</artifactId>
```
Good question. Could be removed I guess.

xhaggi replied:
Ok. I will remove it in a separate commit.

wimdeblauwe commented:
Thank you for this! There are some conflicts now that will need to be resolved. I wanted to release 2.2.0 first to get those changes out before we restructure the whole library.

wimdeblauwe commented:
According to https://docs.github.com/en/repositories/creating-and-managing-repositories/renaming-a-repository we could rename this repository to `htmx-spring-boot` as well as we could be supporting other templating engines in the future as well. WDYT?

checketts commented:
I like the idea of renaming if it is no long Thymeleaf specific

xhaggi commented:
Back to work. Rebased, all conflicts are resolved and the parent and modules are now named `htmx-spring-boot-parent`, `htmx-spring-boot`, `htmx-spring-boot-thymeleaf`. I decided to update the `spring-boot-starter-parent` to the latest version 3.1.2. and therefore I added a new line to the Spring Boot compatibility table in the `README.md`.

wimdeblauwe commented:
> Ah, I didn't know this was a common naming convention for third-party Spring Boot modules. 

I have no idea if it is the "official way", but just a proposal from my side.

xhaggi commented:
It is mentioned in the Spring Boot documentation, but under [Creating Your Own Starter - Naming](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.developing-auto-configuration.custom-starter.naming)
>You should make sure to provide a proper namespace for your starter. Do not start your module names with spring-boot, even if you use a different Maven groupId. We may offer official support for the thing you auto-configure in the future.
As a rule of thumb, you should name a combined module after the starter. For example, assume that you are creating a starter for "acme" and that you name the auto-configure module acme-spring-boot and the starter acme-spring-boot-starter. If you only have one module that combines the two, name it acme-spring-boot-starter.

wimdeblauwe commented on htmx-spring-boot/src/test/resources/templates/fragments.html:
```diff
@@ -2,7 +2,7 @@
 <span th:fragment="active-items-count"
       id="active-items-count"
-      hx:swap-oob="${itemCountSwap} ?: ''"
+      th:hx-swap-oob="${itemCountSwap} ?: ''"
```
Why did you change this?

xhaggi replied:
Since we are not in the thymeleaf module, we cannot use the dialect attributes here.

wimdeblauwe replied:
Ah, right.

wimdeblauwe commented:
@xhaggi Thank you for your work on this. Version 3.0.0 is now released with this split of the modules.

xhaggi commented:
I’m glad I could help with it 😁
